docs/Advanced/Sync.md

```diff
@@ -0,0 +1,163 @@
+- **Implementing `GET changes` API endpoint**
+  - Make sure you perform all queries (and checking for current timestamp) synchronously
+    > This is to ensure that no changes are made to the database while you're fetching changes (otherwise you could never sync some records)
+    - if it's not possible to do so (you have to query each collection separately), be sure to mark `NOW()` to respond with at the *beginning* of the process. You still risk inconsistent responses, but the next pull will fetch whatever changes occured during previous pull.
+- **Implementing `POST changes` API endpoint**
+  - Make sure you perform all changes on all tables in a transaction! If push fails, you don't want partially applied changes.
+  - Compare db record's `last_modified` time with `lastPulledAt`. If it's greater, we have a conflict, and you must abort transaction and return error status.
+  - If client wants to:
+    - … delete a record that don't exist — just ignore it
+    - … update a record that doesn't exist, create it
+    - … create a record that does exist, update it
+  - If there's something wrong with the data format, prefer to "fix" the data if possible instead of failing sync. You don't want the user to have an unsyncable app because of a mistake caused by a bug 5 versions ago.
+  - As with any API, remember to check permissions to create/modify records, make sure you version your API together with local Schema versioning, and all other standard stuff!
+
+### Current limitations
+
+1. If a record being pushed changes between pull and push, push will just fail. It would be better if it failed with a list of conflicts, so that `synchronize()` can automatically respond. Alternatively, sync could only send changed fields and server could automatically always just apply those changed fields to the server version (since that's what per-column client-wins resolver will do anyway)
+2. During next sync pull, changes we've just pushed will be pulled again, which is unnecessary. It would be better if server, during push, also pulled local changes since `lastPulledAt` and responded with NEW timestamp to be treated as `lastPulledAt`.
+3. It shouldn't be necessary to push the whole updated record — just changed fields + ID should be enough
+4. The performance of `synchronize()` has not yet been optimized
+
+### Contributing
+
+1. If you implement Watermelon sync but found this guide confusing, please contribute improvements!
+2. Please help out with solving the current limitations!
+3. If you write server-side code made to be compatible with Watermelon, especially for popular platforms (Node, Ruby on Rails, Kinto, etc.) - please open source it and let us know! This would dramatically simplify implementing sync for people
+4. If you find Watermelon sync bugs, please report the issue! And if possible, write regression tests to make sure it never happens again
+
+## Sync primitives and implementing your own sync
```

Comment:
Maybe it would be better if we split this into 2 separate files? Like if you want to build your own sync engine read this and use this and if you want to try our implementation read this and do this?

Reply:
hmm, it already mostly two files, the bulk of the docs about implementation are here: https://github.com/Nozbe/WatermelonDB/blob/cafe5421981693a0d9b5ca051b3eb271bfc41711/src/sync/index.js#L63 — this is just a reference